### ISSUE TITLE
Tensorflow set_random_seed

### ISSUE THREAD

**Issue by gauss-clb** [tensorflow/tensorflow#15156]
I use `tf.set_random_seed` to set the seed of graph, and run programs many times, but the result is different.
Is it a bug or precision problem?

```python
import tensorflow as tf
slim =tf.contrib.slim

tf.set_random_seed(0)

def vgg16(inputs):
  with slim.arg_scope([slim.conv2d, slim.fully_connected],
                      activation_fn=tf.nn.relu,
                      weights_initializer=tf.truncated_normal_initializer(0.0, 0.01),
                      weights_regularizer=slim.l2_regularizer(0.0005)):
    net = slim.repeat(inputs, 2, slim.conv2d, 64, [3, 3], scope='conv1')
    net = slim.max_pool2d(net, [2, 2], scope='pool1')
    net = slim.repeat(net, 2, slim.conv2d, 128, [3, 3], scope='conv2')
    net = slim.max_pool2d(net, [2, 2], scope='pool2')
    net = slim.repeat(net, 3, slim.conv2d, 256, [3, 3], scope='conv3')
    net = slim.max_pool2d(net, [2, 2], scope='pool3')
    net = slim.repeat(net, 3, slim.conv2d, 512, [3, 3], scope='conv4')
    net = slim.max_pool2d(net, [2, 2], scope='pool4')
    net = slim.repeat(net, 3, slim.conv2d, 512, [3, 3], scope='conv5')
    net = slim.max_pool2d(net, [2, 2], scope='pool5')
    net = slim.flatten(net)
    net = slim.fully_connected(net, 4096, scope='fc6')
    net = slim.dropout(net, 0.5, scope='dropout6')
    net = slim.fully_connected(net, 4096, scope='fc7')
    net = slim.dropout(net, 0.5, scope='dropout7')
    net = slim.fully_connected(net, 1000, activation_fn=None, scope='fc8')
  return net

outputs = vgg16(tf.random_normal([32, 224, 224, 3]))
loss = tf.reduce_mean(tf.square(outputs-tf.random_normal([32, 1000])))
loss = tf.Print(loss, [loss])
optimizer = tf.train.GradientDescentOptimizer(0.001)
train_op = optimizer.minimize(loss)

with tf.Session() as sess:
    tf.global_variables_initializer().run()
    for i in range(10):
        sess.run([loss, train_op])
```

Sometimes it prints
```
2017-12-06 22:12:46.008551: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [1.00103784]
2017-12-06 22:12:48.309682: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [1.01074851]
2017-12-06 22:12:49.039724: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [1.00621808]
2017-12-06 22:12:49.683761: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [1.00629878]
2017-12-06 22:12:50.326798: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [0.999040961]
2017-12-06 22:12:50.973835: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [0.991326749]
2017-12-06 22:12:51.621872: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [1.01098168]
2017-12-06 22:12:52.266909: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [0.989544928]
2017-12-06 22:12:52.911945: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [1.01077247]
2017-12-06 22:12:53.558982: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [1.00654614]
```
Or

```
2017-12-06 22:14:15.614676: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [1.00103784]
2017-12-06 22:14:17.945809: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [1.01074851]
2017-12-06 22:14:18.633848: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [1.00621808]
2017-12-06 22:14:19.283886: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [1.00629878]
2017-12-06 22:14:19.932923: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [0.999040902]
2017-12-06 22:14:20.582960: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [0.99132669]
2017-12-06 22:14:21.231997: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [1.01098168]
2017-12-06 22:14:21.878034: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [0.989544928]
2017-12-06 22:14:22.525071: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [1.01077247]
2017-12-06 22:14:23.137106: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [1.00654614]
```

The fifth and sixth float number are a little different.

If it is the precision problem, the output of program will be different while running on different situation.

If I use` resnet_v1.resnet_v1_101`, the difference becomes bigger.
See below
```
2017-12-06 22:36:10.263869: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.38138771]
2017-12-06 22:36:10.957909: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.37634754]
2017-12-06 22:36:11.511941: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.37431192]
2017-12-06 22:36:12.065973: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.32495236]
2017-12-06 22:36:12.618004: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.31184864]
2017-12-06 22:36:13.171036: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.26968479]
2017-12-06 22:36:13.726067: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.26032281]
2017-12-06 22:36:14.282099: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.21301842]
2017-12-06 22:36:14.831131: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.25101829]
2017-12-06 22:36:15.386162: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.14982963]
```
and
```
017-12-06 22:36:46.220926: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.38138771]
2017-12-06 22:36:46.913966: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.3763473]
2017-12-06 22:36:47.466997: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.37354136]
2017-12-06 22:36:48.022029: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.32750821]
2017-12-06 22:36:48.571060: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.30615568]
2017-12-06 22:36:49.126092: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.26913476]
2017-12-06 22:36:49.682124: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.26520848]
2017-12-06 22:36:50.238156: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.21404171]
2017-12-06 22:36:50.793188: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.25153565]
2017-12-06 22:36:51.347219: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu
\PY\35\tensorflow\core\kernels\logging_ops.cc:79] [3.14605141]
```

**Comment by drpngx** [comment]
did you try to pass the seed to `random_normal`?

**Comment by gauss-clb** [comment]
@drpngx I have tried, but also get the same result, a little difference between different running.

**Comment by drpngx** [comment]
@sguada is there anything special here?

**Comment by gunan** [comment]
There are known issues with op ordering that may prevent you from getting the same results after different runs. Especially with larger graphs this can be a problem.
This is just caused by how the threads are scheduled by the OS.

**Comment by gauss-clb** [comment]
@gunan I find the first number never changes, which means the issues happened during the backpropagation.  So I wonder whether it will cause the gradient updating more unstable.

**Comment by sguada** [comment]
Cuda can also introduce randomness in backpropagation

**Comment by sguada** [comment]
Try passing seed to truncated_normal_initializer and to random_normal

**Comment by gauss-clb** [comment]
@sguada I have tried to pass seed to truncated_normal_initializer and to random_normal, but no effect.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sguada** [comment]
You also need to pass a seed to all tf.random_normal Tensors you create
